config/sample_config_complex.txt
# Sample configuration file for evcouplings complex protein prediction pipeline.
# This file determines all aspects of the computation:
# - which compute environment to use
# - which stages of the pipeline to run
# - what the settings for each of the stages are

# Minimal settings required before this configuration can be executed:
# - set your environment, paths to tools and databases (at the end of this file)
# - under "global", set prefix
# - under "align_1" and "align_2", set the monomer sequence_id 
# - run it! :)

# Configuration rules:
# 1) Global settings override settings for stages
# 2) Outputs of a stage are merged into "global" and fed into the input of subsequent stages
#    (e.g., the alignment_file output of align will be used by the alignment_file input of couplings)
# 3) All settings are explicitly specified here. No hidden defaults in code.
# 4) Each stage is also passed the parameters in the "databases" and "tools" sections

pipeline: protein_complex

# which stages of workflow to run. Uncomment downstream stages using # (however, no stage can be run before the previous
# stage has been run)
stages:
    - align_1
    - align_2
    - concatenate
    - couplings
    - compare
    - mutate
    - fold
    
# Global job settings. These will override settings of the same name in each of the stages.
# These are typically the settings you want to modify for each of your jobs, together with some settings in the align stage.
global:
    # mandatory output prefix of the job (e.g. output/HRAS will store outputs in folder "output", using files prefixed with "HRAS")
    prefix: 

    # Clustering threshold for downweighting redudant sequences (Meff computation). E.g. 0.8 will cluster sequences
    # at a 80% sequence identity cutoff
    theta: 0.9

    # number of cores to use. If running through evcouplings application, will be overriden by environment.cores
    cpu: 2

# Sequence alignment generation/processing for the first monomer.
align_1:
    # use complex protocol to properly prepare inputs for concatenation
    protocol: complex
    
    # monomer alignment creation protocol to nest within the complex alignment protocol
    # choose either existing (below) to use a previously created alignment
    # or standard to construct an alignment 
    alignment_protocol: standard


    # Mandatory: specify the sequence identifier
    # Region can be left blank
    # Sequence file can be left blank
    sequence_id: 
    region: 
    sequence_file:
    
    # The following typically do not need to be set because 'global' overrides them
    # prefix:
    # theta:

    # index of first residue in sequence_id / sequence_file. This can be used to renumber sequences that already have
    # been cut to a subsequence
    first_index: 1

    # Use bitscore threshold instead of E-value threshold for sequence search
    use_bitscores: True

    # jackhmmer domain- and sequence-level inclusion thresholds.
    # if use_bitscores is True:
    # - floating point number will be interpreted as a relative bitscore threshold (bits/residue)
    # - integer will be interpreted as an absolute bitscore threshold
    # if use_bitscore is False:
    # - mantissa-exponent string or float will be interpreted literally
    # - integer will be interpreted as negative of the exponent (10 -> 1E-10)
    domain_threshold: 0.5
    sequence_threshold: 0.5

    # number of jackhmmer iterations
    iterations: 5

    # sequence database (specify possible databases and paths in "databases" section below)
    # note: use uniprot for genome distance based concatenation
    database: uniref100

    # compute the redundancy-reduced number of effective sequences (M_eff) already in the alignment stage.
    # To save compute time, this computation is normally carried out in the couplings stage
    compute_num_effective_seqs: True

    # Filter sequence alignment at this % sequence identity cutoff. Can be used to cut computation time in
    # the couplings stage (e.g. set to 95 to remove any sequence that is more than 95% identical to a sequence
    # already present in the alignment). If blank, no filtering. If filtering, HHfilter must be installed.
    seqid_filter:

    # Only include alignment columns with at least x% residues (rather than gaps) during model inference
    minimum_sequence_coverage: 50

    # Only keep sequences that align to at least x% of the target sequence (i.e. remove fragments)
    minimum_column_coverage: 70

    # Create a file with extracted annotation from UniRef/UniProt sequence FASTA headers
    extract_annotation: True
    cpu:

    # set to True to turn of jackhmmer bias correction
    nobias: False

    # if align stage has been run previously, reuse the generated raw sequence alignment coming out of jackhmmer
    reuse_alignment: True

    # create checkpoint files of HMM and aligment after each iteration
    checkpoints_hmm: False
    checkpoints_ali: False

# Alternative protocol: reuse existing alignment and apply postprocessing to generate alignment that is consistent
# with pipeline requirements. Uncomment, and comment all values in align section above to enable the "existing" protocol
#    protocol: complex
#    alignment_protocol: existing
#    prefix:
#    # Path of input alignment. Alignment needs to contain region in form SEQID/start-end, or first_index must be set
#    input_alignment:
#    sequence_id:
#    first_index:
#    compute_num_effective_seqs: False
#    theta:
#    seqid_filter:
#    minimum_sequence_coverage: 50
#    minimum_column_coverage: 70
#    extract_annotation: True

#    # if using existing alignment protocol, provide a path to the annotations.csv file 
#    # from the monomer run that generated the input alignment
#    # Needed to correctly find the species identifiers for best hit concatenation
#    override_annotation_file: 

# Sequence alignment generation/processing for the second monomer.
align_2:
    # use complex protocol to properly prepare inputs for concatenation
    protocol: complex


    alignment_protocol: standard
    # Mandatory: specify the sequence identifier and region
    # Sequence file can be left blank
    sequence_id: 
    region:
    sequence_file:
    
    # The following typically do not need to be set because 'global' overrides them
    # prefix:
    # theta:

    # index of first residue in sequence_id / sequence_file. This can be used to renumber sequences that already have
    # been cut to a subsequence
    first_index: 1

    # Use bitscore threshold instead of E-value threshold for sequence search
    use_bitscores: True

    # jackhmmer domain- and sequence-level inclusion thresholds.
    # if use_bitscores is True:
    # - floating point number will be interpreted as a relative bitscore threshold (bits/residue)
    # - integer will be interpreted as an absolute bitscore threshold
    # if use_bitscore is False:
    # - mantissa-exponent string or float will be interpreted literally
    # - integer will be interpreted as negative of the exponent (10 -> 1E-10)
    domain_threshold: 0.5
    sequence_threshold: 0.5

    # number of jackhmmer iterations
    iterations: 5

    # sequence database (specify possible databases and paths in "databases" section below)
    # note: use uniprot for genome distance based concatenation
    database: uniref100

    # compute the redundancy-reduced number of effective sequences (M_eff) already in the alignment stage.
    # To save compute time, this computation is normally carried out in the couplings stage
    compute_num_effective_seqs: True

    # Filter sequence alignment at this % sequence identity cutoff. Can be used to cut computation time in
    # the couplings stage (e.g. set to 95 to remove any sequence that is more than 95% identical to a sequence
    # already present in the alignment). If blank, no filtering. If filtering, HHfilter must be installed.
    seqid_filter:

    # Only include alignment columns with at least x% residues (rather than gaps) during model inference
    minimum_sequence_coverage: 50

    # Only keep sequences that align to at least x% of the target sequence (i.e. remove fragments)
    minimum_column_coverage: 70

    # Create a file with extracted annotation from UniRef/UniProt sequence FASTA headers
    extract_annotation: True
    cpu:

    # set to True to turn of jackhmmer bias correction
    nobias: False

    # if align stage has been run previously, reuse the generated raw sequence alignment coming out of jackhmmer
    reuse_alignment: True

    # create checkpoint files of HMM and aligment after each iteration
    checkpoints_hmm: False
    checkpoints_ali: False

# Alternative protocol: reuse existing alignment and apply postprocessing to generate alignment that is consistent
# with pipeline requirements. Uncomment, and comment all values in align section above to enable the "existing" protocol
#    protocol: complex
#    alignment_protocol: existing
#    prefix:
#    # Path of input alignment. Alignment needs to contain region in form SEQID/start-end, or first_index must be set
#    input_alignment:
#    sequence_id:
#    first_index:
#    compute_num_effective_seqs: False
#    theta:
#    seqid_filter:
#    minimum_sequence_coverage: 50
#    minimum_column_coverage: 70
#    extract_annotation: True
#    # if using existing alignment protocol, provide a path to the annotations.csv file 
#    # from the monomer run that generated the input alignment
#    # Needed to correctly find the species identifiers for best hit concatenation
#    override_annotation_file: 

#Generation of concatenated sequence alignment for evolutionary couplings calculation
concatenate:

    # Alignment files to use in concatenation
    # Will be usually overriden by global settings / output of previous stage
    first_alignment_file:
    second_alignment_file:

    # Select protocol for concatenation of sequence alignments
    # Available protocols: 
    # genome_distance: pair sequences that are closest neighbors on the genome
    # best_hit: for each genome, pair the sequences that have the highest % identity to the target sequence
    # for best hit protocol, user can set use_best_reciprocal to take the best reciprocal hits only (recommended)
    # inter_species: for each genome, pair the sequences that have the highest % identity to the target sequence
    # for inter_species protocol, user should define a path for the species-species file for truthful
    # concatenation TODO
    protocol: best_hit
    use_best_reciprocal: true
    # species_species_file: <path_to_species_species_file>
    
    # Maximum genome distance in bases allowed between pairs
    # Required for genome_distance protocol only
    genome_distance_threshold: 10000
    
    # Maximum sequence identity allowed for hits to be designated
    # as paralogs. Required for best_hit in best reciprocal mode only
    paralog_identity_threshold: 0.95
    
    # Parameters for filtering of concatenated alignment
    
    # Filter sequence alignment at this % sequence identity cutoff. Can be used to cut computation time in
    # the couplings stage (e.g. set to 95 to remove any sequence that is more than 95% identical to a sequence
    # already present in the alignment). If blank, no filtering. If filtering, HHfilter must be installed.
    seqid_filter:
    
    # Only keep sequences that align to at least x% of the target sequence (i.e. remove fragments)
    minimum_sequence_coverage: 50
    
    # Only include alignment columns with at least x% residues (rather than gaps) during model inference
    minimum_column_coverage: 50

    # compute the redundancy-reduced number of effective sequences (M_eff) already in the alignment stage.
    # To save compute time, this computation is normally carried out in the couplings stage
    compute_num_effective_seqs: False
    # typically does not need to be set as 'global' overrides
    theta:

# Inference of evolutionary couplings from sequence alignment
couplings:
    # current options: standard (model inference using plmc, postprocessing for monomers)
    # complex (model inference using plmc, postprocessing for complexes)
    protocol: complex

    # number of plmc iterations
    iterations: 100

    # specify custom alphabet as a string. Gap symbol must be first character
    alphabet:

    # Treat gaps as missing data during model inference
    ignore_gaps: True

    # strength of regularization on coupling parameters J
    lambda_J: 0.01

    # adjust for larger number of coupling parameters relative to number of fields h (multiply by model length and
    # number of states)
    lambda_J_times_Lq: True

    # strength of regularization on fields h
    lambda_h: 0.01
    lambda_group:
    scale_clusters:

    # reuse ECs and model parameters, if this stage has been run before
    reuse_ecs: True

    # Sequence separation filter for generation of CouplingScores_longrange.csv table (i.e. to take out short-range
    # ECs from table, only pairs with abs(i-j)>=min_sequence_distance will be kept.
    min_sequence_distance: 6
    
    # Parameters specific to complex pipeline scoring
    # Scoring model to assess confidence in computed ECs
    # available options: skewnormal, normal, evcomplex
    scoring_model: skewnormal
    
    # Specify whether to use all ECs or only inter-molecular ECs for scoring
    use_all_ecs_for_scoring: False

    # Following input parameters will usually be overriden by "global" and outputs of "align" stage
    # prefix:
    # alignment_file:
    # focus_sequence:
    focus_mode: True
    # segments:
    # cpu:
    # theta:

# Compare ECs to known 3D structures
compare:
    # Current options: standard, complex
    protocol: complex
    
    # Following parameters will be usually overriden by global settings / output of previous stage
    prefix:
    ec_file:
    first_sequence_id:
    first_target_sequence_file:
    second_sequence_id:
    second_target_sequence_file:

    # If True, find structures by sequence alignment against the PDB, otherwise identify structures using
    # sequence_id and SIFTS database (sequence_id must be UniProt AC/ID in this case)
    first_by_alignment: True
    second_by_alignment: True
    # Alignment method to use to search the PDB Seqres database. Options: jackhmmer, hmmsearch
    # Set to jackhmmer to search the PDB Seqres database using jackhmmer from the target sequence only (more stringent). 
    # Set to hmmsearch to search the PDB seqres database using an HMM built from the output monomer alignment (less stringent). 
    # Warning: searching by HMM may result in crystal structures from very distant homologs or even unrelated sequences. 
    first_pdb_alignment_method: jackhmmer
    second_pdb_alignment_method: jackhmmer
    
    # Leave this parameter empty to use all PDB structures for given sequence_id, otherwise
    # will be limited to the given IDs (single value or list). Important: note that this acts only as a filter on the
    # structures found by alignment or in the SIFTS table (!)
    pdb_ids:
    first_pdb_ids:
    second_pdb_ids:

    # Limit number of structures and chains for comparison
    # Note - the intersection of the monomer structural hits is taken to find the
    # Inter-protein structures. If you limit the number of monomer structures found in this step, 
    # you may miss some inter-protein structures
    first_max_num_structures: 100
    first_max_num_hits: 100
    second_max_num_structures: 100
    second_max_num_hits: 100

    # compare to multimer contacts (if multiple chains of the same sequence or its homologs are present in a structure)
    first_compare_multimer: True
    second_compare_multimer: True

    # settings for sequence alignment against PDB sequences using jackhmmer
    # (additional settings like iterations possible, compare to align stage)
    first_sequence_file:
    first_first_index:
    first_region:
    first_alignment_min_overlap: 20
    first_use_bitscores: True
    first_domain_threshold: 0.5
    first_sequence_threshold: 0.5
    
    second_sequence_file:
    second_first_index:
    second_region:
    second_alignment_min_overlap: 20
    second_use_bitscores: True
    second_domain_threshold: 0.5
    second_sequence_threshold: 0.5

    # Comparison and plotting settings
    
    # Return an error if we fail to automatically retrieve information about a given pdb id
    raise_missing: False
    
    # Filter that defines which atoms will be used for distance calculations. If empty/None, no filter will be
    # applied (resulting in the computation of minimum atom distances between all pairs of atoms). If setting to any
    # particular PDB atom type, only these atoms will be used for the computation (e.g. CA will give C_alpha distances,
    # CB will give C_beta distances, etc.)
    atom_filter:

    # Distance cutoff (Angstrom) for a true positive pair
    distance_cutoff: 5

    # Only long-range pairs with abs(i-j)>= min_sequence_distance will be used for CouplingScoresCompared_longrange.csv file
    min_sequence_distance: 6

    # Plot contact maps with ECs above these mixture model probability cutoffs
    plot_probability_cutoffs: [0.90, 0.99]

    # Plot fixed numbers of inter-protein ECS, and all intra ECs scoring at least as high
    # As those inter-protein ECs. 
    # Use integers only
    plot_lowest_count: 5
    plot_highest_count: 10
    plot_increase: 5

    # Axis boundaries of contact map plot depending on range of ECs and structure.
    # Options: union, intersection, ecs, structure, [start, end] (e.g. [100, 200])
    boundaries: ecs

    # scale sizes of EC dots in scatter plot based on strength of EC score
    scale_sizes: True

    # draw secondary structure on contact map plots
    draw_secondary_structure: True
    
# Settings for Mutation effect predictions
mutate:
    # Options: standard, complex
    protocol: complex

    # predict the following dataset file (.csv file, mutants like A102V or A102V,K199W in column "mutant"
    # segments in column "segment" as "A_1" and "B_1" for first and second monomers, one segment specified per mutation)
    mutation_dataset_file:

    # Inputs set by global stage and output of previous stages
    # prefix:
    # model_file:

# Settings for generating restraints for docking (using Haddock)
fold:
    protocol: complex_dock

    # Inputs set by global stage and output of previous stages
    prefix:

    # generate restraint files with inter-protein ECs above the following score cutoffs
    dock_probability_cutoffs:
      - 0.9
      - 0.99

    # generate restraint files fixed numbers of ECS. Integers will be interpreted as absolute numbers,
    # floats as fractions of L (model length)
    dock_lowest_count: 5
    dock_highest_count: 10
    dock_increase: 5


# These settings allow job status tracking using a database, and result collection in an archive
management:
    # URI of database
    database_uri:

    # unique job identifier
    job_name:

    # add the following output files to results archive
    archive: [target_sequence_file, statistics_file, alignment_file, frequencies_file, ec_file, ec_longrange_file,
              model_file, enrichment_file, evzoom_file, enrichment_pml_files, ec_lines_pml_file, contact_map_files,
              ec_compared_all_file, ec_compared_longrange_file, remapped_pdb_files, mutations_epistatic_pml_files,
              mutation_matrix_file, mutation_matrix_plot_files, secondary_structure_pml_file, folding_ec_file,
              folded_structure_files, folding_ranking_file, folding_comparison_file, folding_individual_comparison_files,
              ec_lines_compared_pml_file, pdb_structure_hits_file, sec_struct_file]

    # Delete the following output files after running the job if you don't need them, to save disk space.
    # Note that this may jeopardize your ability to rerun parts of the job if intermediate files are missing.
    # The following, deactivated default deletes the biggest output files.
    # delete: [raw_alignment_file, model_file]

# Computational environment for batch jobs (using evcouplings command line application)
environment:
    # current options for engine: lsf, slurm, local (for local, only set cores and leave all other fields blank)
    # If your batch engine of choice (e.g. SGE, Torque) is not available yet, please consider contributing by
    # implementing it and submitting a pull request!
    # Note that "cores" will override the "cpu" parameter for "global"
    engine: slurm
    queue: medium
    cores: 2
    memory: 15000
    time: 2-0:0:0

    # Special setting for "local" engine to define number of workers running in parallel
    # (note that "cores" has to be defined above to make sure each job only uses a defined
    # number of cores). If not defined or None, will default to number of cores / cores per job;
    # otherwise specify integer to limit number of workers (1 for serial execution of subjobs)
    # parallel_workers: 1

    # command that will be executed before running actual computation (can be used to set up environment)
    configuration:
        

# Paths to databases used by evcouplings.
databases:
    # Sequence databases (only download the ones you want to use). You can also specify arbitrary databases in FASTA format
    # using a database name of your choice here)
    uniprot: /n/groups/marks/databases/jackhmmer/uniprot/uniprot_current.o2.fasta
    uniref100: /n/groups/marks/databases/jackhmmer/uniref100/uniref100_current.o2.fasta
    uniref90: /n/groups/marks/databases/jackhmmer/uniref90/uniref90_current.o2.fasta

    # URL do download sequences if sequence_file is not given. {} will be replaced by sequence_id.
    sequence_download_url: http://rest.uniprot.org/uniprot/{}.fasta

    # Directory with PDB MMTF structures (leave blank to fetch structures from web)
    pdb_mmtf_dir:

    # SIFTS mapping information. Point to file paths in an existing directory, and if these files do not exist, they will be
    # automatically generated and saved at the given file path (this may take a while).
    # Periodically delete these files to more recent versions of SIFTS are used.
    sifts_mapping_table: /n/groups/marks/databases/SIFTS/pdb_chain_uniprot_plus_current.o2.csv
    sifts_sequence_db: /n/groups/marks/databases/SIFTS/pdb_chain_uniprot_plus_current.o2.fasta
    
    # the following two databases are exclusive to EVcomplex and need to be manually downloaded and saved locally
    # then add the paths to your local copies of the database
    # Download urls: 
    # ena_genome_location_table: https://marks.hms.harvard.edu/evcomplex_databases/cds_pro_2017_02.txt
    # uniprot_to_embl_table: https://marks.hms.harvard.edu/evcomplex_databases/idmapping_uniprot_embl_2017_02.txt 
    uniprot_to_embl_table: /n/groups/marks/databases/complexes/idmapping/idmapping_uniprot_embl_2017_02.txt
    ena_genome_location_table: /n/groups/marks/databases/complexes/ena/2017_02/cds_pro.txt

# Paths to external tools used by evcouplings. Please refer to README.md for installation instructions and which tools are required.
tools:
    jackhmmer: /n/groups/marks/pipelines/evcouplings/software/hmmer-3.1b2-linux-intel-x86_64/binaries/jackhmmer
    hmmbuild: /n/groups/marks/pipelines/evcouplings/software/hmmer-3.1b2-linux-intel-x86_64/binaries/hmmbuild
    hmmsearch: /n/groups/marks/pipelines/evcouplings/software/hmmer-3.1b2-linux-intel-x86_64/binaries/hmmsearch
    plmc: /n/groups/marks/pipelines/evcouplings/software/plmc/bin/plmc
    hhfilter: /n/groups/marks/pipelines/evcouplings/software/hh-suite/bin/hhfilter
    psipred: /n/groups/marks/software/runpsipred
    cns: /n/groups/marks/pipelines/evcouplings/software/cns_solve_1.21/intel-x86_64bit-linux/bin/cns
    maxcluster: /n/groups/marks/pipelines/evcouplings/software/maxcluster64bit
    


evcouplings/visualize/parameters.py
"""
Visualization of pair model parameters

Authors:
  Thomas A. Hopf
"""

import json
import numpy as np
from evcouplings.couplings.pairs import add_mixture_probability


def evzoom_data(
    model,
    ec_threshold=0.9,
    freq_threshold=0.01,
    Jij_threshold=10,
    score="cn",
    reorder=None,
):
    """
    Generate data for EVzoom visualization. Use evzoom_json()
    to get final JSON string to use with EVzoom.

    Parameters
    ----------
    model : evcouplings.couplings.model.CouplingsModel
        Parameters of pairwise graphical model
    ec_threshold : float or int, optional (default: 0.9)
        Only display evolutionary couplings above this
        threshold. If float between 0 and 1, this will be
        interpreted as probability cutoff for mixture model.
        Otherwise, will be interpreted as absolute number of couplings.
    freq_threshold : float, optional (default: 0.01)
        Only display coupling parameters for amino acids with
        at least this frequency in the underlying sequence
        alignment
    Jij_threshold : int or float, optional (default: 10)
        Only display coupling parameters above this
        threshold. If float, this number will be interpreted
        as an actual score threshold; if int, this will
        be interpreted as a percentage of the maximum
        absolute score.
    score : str, optional (default: "cn")
        Use this score to determine which couplings to display.
        Valid choices are the score columns contained in the
        CouplingsModel.ecs dataframe
    reorder : str, optional (default: "KRHEDNQTSCGAVLIMPYFW")
        Order of amino acids in displayed coupling matrices

    Returns
    -------
    map_ : dict
        Map containing sequence indices and characters
    logo : list
        List containing information about sequence logos
        for axes of visualization
    matrix : dict
        List containing couplings that will be visualized
    """
    DIGITS = 2
    DIGITS_LOGO = 2
    ecs = model.ecs

    if 0 < ec_threshold <= 1.0:
        ecs = add_mixture_probability(ecs, score=score)
        ecs_sel = ecs.loc[ecs.probability >= ec_threshold]
    else:
        ecs_sel = ecs.iloc[: int(ec_threshold)]

    # if cutoff for couplings is given as int, interpret
    # as percentage of biggest absolute coupling value
    if isinstance(Jij_threshold, int):
        max_val = np.max(np.abs(model.Jij()))
        Jij_threshold = max_val * Jij_threshold / 100

    if reorder is not None:
        alphabet = np.array(list(reorder))
        alphabet_order = [model.alphabet_map[c] for c in reorder]
    else:
        alphabet = model.alphabet
        alphabet_order = sorted(model.alphabet_map.values())

    # Map containing sequence and indeces
    map_ = {
        "letters": "".join(model.seq()),
        "indices": list(map(int, model.sn())),
    }

    # assemble coupling matrix
    matrix = []

    for idx, r in ecs_sel.iterrows():
        i, j, score_ij = r["i"], r["j"], r[score]
        Jij = model.Jij(i, j)[alphabet_order, :][:, alphabet_order]
        ai_set = np.where(np.max(np.abs(Jij), axis=1) > Jij_threshold)[0]
        aj_set = np.where(np.max(np.abs(Jij), axis=0) > Jij_threshold)[0]

        cur_matrix = [
            [round(Jij[ai, aj], DIGITS) for aj in list(aj_set)] for ai in list(ai_set)
        ]

        cur_matrix_T = [
            [round(Jij[ai, aj], DIGITS) for ai in list(ai_set)] for aj in list(aj_set)
        ]

        cur_row = {
            "i": model.mn(i) + 1,
            "j": model.mn(j) + 1,
            "score": round(score_ij, DIGITS),
            "iC": "".join(alphabet[ai_set]),
            "jC": "".join(alphabet[aj_set]),
            "matrix": cur_matrix,
        }

        cur_row_T = {
            "i": cur_row["j"],
            "j": cur_row["i"],
            "score": cur_row["score"],
            "iC": cur_row["jC"],
            "jC": cur_row["iC"],
            "matrix": cur_matrix_T,
        }

        matrix.append(cur_row)
        matrix.append(cur_row_T)

    # assemble sequence logo
    fi = model.fi()
    q = model.num_symbols

    B = -fi * np.log2(fi, where=fi > 0)
    B[fi <= 0] = 0
    R = np.log2(q) - B.sum(axis=1)

    logo = []
    for i in range(model.L):
        order = np.argsort(fi[i, :])
        frequent = order[fi[i, order] >= freq_threshold]

        symbols = model.alphabet[frequent]
        fi_row = fi[i, frequent] * R[i]

        logo.append(
            [
                {"code": s, "bits": round(float(h), DIGITS_LOGO)}
                for s, h in zip(symbols, fi_row)
            ]
        )

    return map_, logo, matrix


def evzoom_json(model, **kwargs):
    """
    Create JSON input for visualizing parameters
    of pairwise graphical model using EVzoom

    Parameters
    ----------
    model : evcouplings.couplings.model.CouplingsModel
        Parameters of pairwise graphical model
    **kwargs
        See evzoom_data() for options

    Returns
    -------
    str
        EVzoom-ready JSON input
    """
    map_, logo, matrix = evzoom_data(model, **kwargs)

    data = {
        "map": map_,
        "logo": logo,
        "couplings": matrix,
    }

    return json.dumps(data)
